### PR TITLE
Recursively create partitioned indexes

### PR DESCRIPTION
**Pull request by kalensk**
Greenplum does not enforce constraint and index names to match which differs from upstream behavior. This presents a challenge when upgrading. Specifically in Greenplum, pg_dump’s generated DDL's are not specific enough to reproduce the old database catalog. See [proposal](https://docs.google.com/document/d/1E_pFWbzWOgBgHQIXX6yuOxZqMtTnpKbE51UUiAJaEI4/) for details.

This PR is a first step towards enforcing constraint and index names to match in order to align with upstream and enable upgrading of Greenplum databases.

Pull from upstream Postgres to make DefineIndex recursively create partitioned indexes. Instead of creating an individual IndexStmt for every partition, create indexes by recursing on the partition children.  This aligns index creation with upstream in preparation for adding INTERNAL_AUTO relationships between partition indexes.

 * The QD will now choose the same name for partition indexes as Postgres.
 * Update tests to reflect the partition index names changes.
 * The changes to DefineIndex are mostly cherry-picked from Postgres commit: 8b08f7d4820
 * transformIndexStmt and its callers have been aligned with Postgres REL9_4_STABLE

We also needed to revert some recent changes to dispatch, please see the associated commit message for details.

### CODE REVIEW

File: src/backend/commands/indexcmds.c [hunk 1]
```diff
@@ -807,6 +836,121 @@ DefineIndex(Oid relationId,
...
+	{
+		/*
+		 * Unless caller specified to skip this step (via ONLY), process each
+		 * partition to make sure they all contain a corresponding index.
+		 *
+		 * GPDB: Upstream uses stmt->relation here to determine whether to continue
+		 * recusing in DefineIndex. However, GPDB always sets stmt->relation
+		 * because it needs to be dispatched to the QEs.
+		 */
+		if (interpretInhOption(stmt->relation->inhOpt) && (Gp_role == GP_ROLE_DISPATCH))
```
[the remark this answers is not in the document]
Comment (tvesely) [reply src/backend/commands/indexcmds.c]: This is the only place that I have found (so far) where that flag would be useful, and this works relatively well. If I find another place that could benefit I'll add a field.

File: src/test/isolation2/expected/reindex/reindextable_while_altertable_ao_part_btree.out [hunk 2]
```diff
@@ -13,21 +13,21 @@ INSERT 21
 Insert into reindex_crtabforalter_part_ao_btree select i, to_date('2013-11-'||i, 'YYYY/MM/DD') , 29.51+i from generate_series(10,30) i;
 INSERT 21
 
-create index idx_reindex_crtabforalter_part_ao_btree on reindex_crtabforalter_part_ao_btree(id);
+create index on reindex_crtabforalter_part_ao_btree(id);
```
Comment (jchampio) [inline src/test/isolation2/expected/reindex/reindextable_while_altertable_ao_part_btree.out]: The naming scheme for automatically created indexes has changed, but there have been quite a few removals of the manual index naming as well. Is manually naming indexes no longer supported, or is there another reason to make these changes to the tests?

Comment (tvesely) [reply src/test/isolation2/expected/reindex/reindextable_while_altertable_ao_part_btree.out]: They were manually naming the index to the same name that the index would have automatically generated anyway, so we just removed the name. By doing it this way, if the index naming scheme changes in upstream again, we won't need to modify these particular commands in the tests, and will only need to change the commands that follow.

File: src/backend/commands/indexcmds.c [hunk 3]
```diff
@@ -807,6 +829,119 @@ DefineIndex(Oid relationId,
...
```
Comment (berlin-ab) [inline src/backend/commands/indexcmds.c]: What does `interpretInhOption` interpret? What question is it trying to answer & that's what the name of the function should probably be.

File: src/backend/commands/indexcmds.c [hunk 4]
```diff
@@ -425,16 +429,10 @@ DefineIndex(Oid relationId,
 	LOCKMODE	lockmode;
 	Snapshot	snapshot;
 	bool		need_longlock = true;
-	bool		shouldDispatch;
+	bool		shouldDispatch = Gp_role == GP_ROLE_DISPATCH && !IsBootstrapProcessingMode();
```
Comment (berlin-ab) [inline src/backend/commands/indexcmds.c]: While I personally don't mind this style of variable declaration, it seems that we tend to declare a variable in one place, and then initialize it in another.

File: src/backend/commands/indexcmds.c (hunk 3 above)
Comment (berlin-ab) [reply src/backend/commands/indexcmds.c]: It seems that this function comes from upstream. That doesn't stop us from creating a better named passthrough macro or function.

File: src/backend/commands/indexcmds.c (hunk 4 above)
Comment (tvesely) [reply src/backend/commands/indexcmds.c]: Agreed. Fixed.

**Comment by hlinnaka** [comment]
With this commit, you no longer get these "building index for child partition" NOTICEs:
```
postgres=# create table foo (id int) partition by range (id) (start (1) end(2));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'id' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "foo_1_prt_1" for table "foo"
CREATE TABLE
postgres=# create index on foo (id);
NOTICE:  building index for child partition "foo_1_prt_1"
CREATE INDEX
```
Was that intentional? If it was, please strip them away from the expected outputs, too. They're not causing regression failures, because all NOTICEs are ignored when comparing the outputs, but it would be nice to keep the expected outputs in sync with what we actually print, anyway. If it was not intentional, please bring back the notice.

**Comment by dkrieger-pvtl** [comment]
> With this commit, you no longer get these "building index for child partition" NOTICEs:
> 
> Was that intentional? If it was, please strip them away from the expected outputs, too. They're not causing regression failures, because all NOTICEs are ignored when comparing the outputs, but it would be nice to keep the expected outputs in sync with what we actually print, anyway. If it was not intentional, please bring back the notice.

Yes, it was intentional as upstream apparently does not do so for indexes on partitions.  We created a note to ourselves to fix this as a chore.

**Pull request merged**